File: services/api-gateway/src/routes/health.ts
// Health check endpoints for API Gateway
import { Router } from 'express';
import Redis from 'ioredis';
import { config } from '../config';

const router = Router();
const redis = new Redis(config.redisUrl);

// Liveness probe - is the service running?
router.get('/live', (req, res) => {
  res.status(200).json({
    status: 'ok',
    service: 'api-gateway',
    timestamp: new Date().toISOString(),
  });
});

// Readiness probe - is the service ready to accept traffic?
router.get('/ready', async (req, res) => {
  try {
    // Check Redis connectivity
    await redis.ping();

    res.status(200).json({
      status: 'ready',
      service: 'api-gateway',
      checks: {
        redis: 'ok',
      },
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    res.status(503).json({
      status: 'not_ready',
      service: 'api-gateway',
      checks: {
        redis: 'failed',
      },
      error: error instanceof Error ? error.message : 'Unknown error',
      timestamp: new Date().toISOString(),
    });
  }
});

// Aggregated health check for all downstream services
router.get('/', async (req, res) => {
  // TODO: Implement health aggregation for all services
  res.status(200).json({
    status: 'ok',
    service: 'api-gateway',
    version: '1.0.0',
    timestamp: new Date().toISOString(),
  });
});

export { router as healthRouter };

File: services/catalog-service/src/api/routes/health.py
"""Health check endpoints for catalog service"""
from fastapi import APIRouter, status
from pydantic import BaseModel
from datetime import datetime

from ..infrastructure.database import get_database
from ..infrastructure.search import get_opensearch_client

router = APIRouter()


class HealthResponse(BaseModel):
    """Health check response"""
    status: str
    service: str
    version: str
    timestamp: datetime
    checks: dict


@router.get("/live", status_code=status.HTTP_200_OK)
async def liveness():
    """Liveness probe - is the service running?"""
    return {
        "status": "ok",
        "service": "catalog-service",
        "timestamp": datetime.utcnow().isoformat(),
    }


@router.get("/ready", response_model=HealthResponse)
async def readiness():
    """Readiness probe - is the service ready to accept traffic?"""
    checks = {}

    # Check MongoDB
    try:
        db = get_database()
        await db.command("ping")
        checks["mongodb"] = "ok"
    except Exception as e:
        checks["mongodb"] = f"failed: {str(e)}"

    # Check OpenSearch
    try:
        es = get_opensearch_client()
        await es.ping()
        checks["opensearch"] = "ok"
    except Exception as e:
        checks["opensearch"] = f"failed: {str(e)}"

    # Determine overall status
    all_ok = all(v == "ok" for v in checks.values())

    return HealthResponse(
        status="ready" if all_ok else "not_ready",
        service="catalog-service",
        version="1.0.0",
        timestamp=datetime.utcnow(),
        checks=checks,
    )


@router.get("/", response_model=HealthResponse)
async def health():
    """Aggregated health check"""
    return await readiness()
